File: apps/api/src/main/java/com/ecommerce/inventory/service/ScheduledTaskRegistry.java
package com.ecommerce.inventory.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Registry service for managing and monitoring scheduled tasks.
 * Provides capabilities to track task execution, failures, and performance metrics.
 */
@Service
public class ScheduledTaskRegistry {
    
    private static final Logger logger = LoggerFactory.getLogger(ScheduledTaskRegistry.class);
    
    private final Map<String, TaskExecutionInfo> taskRegistry = new ConcurrentHashMap<>();
    private final Map<String, TaskMetrics> taskMetrics = new ConcurrentHashMap<>();
    
    /**
     * Register a scheduled task for monitoring.
     */
    public void registerTask(String taskName, String description) {
        TaskExecutionInfo info = new TaskExecutionInfo(taskName, description);
        taskRegistry.put(taskName, info);
        taskMetrics.put(taskName, new TaskMetrics());
        
        logger.info("Registered scheduled task: {} - {}", taskName, description);
    }
    
    /**
     * Record the start of a task execution.
     */
    public void recordTaskStart(String taskName) {
        TaskExecutionInfo info = taskRegistry.get(taskName);
        if (info != null) {
            info.setLastExecutionStart(LocalDateTime.now());
            info.setCurrentlyExecuting(true);
            
            TaskMetrics metrics = taskMetrics.get(taskName);
            if (metrics != null) {
                metrics.incrementExecutionCount();
            }
            
            logger.debug("Task started: {}", taskName);
        }
    }
    
    /**
     * Record the successful completion of a task execution.
     */
    public void recordTaskSuccess(String taskName, long executionTimeMs) {
        TaskExecutionInfo info = taskRegistry.get(taskName);
        if (info != null) {
            info.setLastExecutionEnd(LocalDateTime.now());
            info.setLastExecutionDuration(executionTimeMs);
            info.setCurrentlyExecuting(false);
            info.setLastExecutionStatus("SUCCESS");
            
            TaskMetrics metrics = taskMetrics.get(taskName);
            if (metrics != null) {
                metrics.incrementSuccessCount();
                metrics.updateAverageExecutionTime(executionTimeMs);
                metrics.updateMaxExecutionTime(executionTimeMs);
            }
            
            logger.debug("Task completed successfully: {} ({}ms)", taskName, executionTimeMs);
        }
    }
    
    /**
     * Record a task execution failure.
     */
    public void recordTaskFailure(String taskName, long executionTimeMs, String errorMessage) {
        TaskExecutionInfo info = taskRegistry.get(taskName);
        if (info != null) {
            info.setLastExecutionEnd(LocalDateTime.now());
            info.setLastExecutionDuration(executionTimeMs);
            info.setCurrentlyExecuting(false);
            info.setLastExecutionStatus("FAILED");
            info.setLastErrorMessage(errorMessage);
            
            TaskMetrics metrics = taskMetrics.get(taskName);
            if (metrics != null) {
                metrics.incrementFailureCount();
                metrics.updateAverageExecutionTime(executionTimeMs);
            }
            
            logger.warn("Task failed: {} ({}ms) - {}", taskName, executionTimeMs, errorMessage);
        }
    }
    
    /**
     * Get execution information for a specific task.
     */
    public TaskExecutionInfo getTaskInfo(String taskName) {
        return taskRegistry.get(taskName);
    }
    
    /**
     * Get metrics for a specific task.
     */
    public TaskMetrics getTaskMetrics(String taskName) {
        return taskMetrics.get(taskName);
    }
    
    /**
     * Get all registered tasks.
     */
    public Map<String, TaskExecutionInfo> getAllTasks() {
        return new ConcurrentHashMap<>(taskRegistry);
    }
    
    /**
     * Get metrics for all tasks.
     */
    public Map<String, TaskMetrics> getAllMetrics() {
        return new ConcurrentHashMap<>(taskMetrics);
    }
    
    /**
     * Check if any tasks are currently executing.
     */
    public boolean hasRunningTasks() {
        return taskRegistry.values().stream()
                .anyMatch(TaskExecutionInfo::isCurrentlyExecuting);
    }
    
    /**
     * Get count of currently executing tasks.
     */
    public long getRunningTaskCount() {
        return taskRegistry.values().stream()
                .mapToLong(info -> info.isCurrentlyExecuting() ? 1 : 0)
                .sum();
    }
    
    /**
     * Information about a scheduled task execution.
     */
    public static class TaskExecutionInfo {
        private final String taskName;
        private final String description;
        private final LocalDateTime registeredAt;
        private LocalDateTime lastExecutionStart;
        private LocalDateTime lastExecutionEnd;
        private long lastExecutionDuration;
        private boolean currentlyExecuting;
        private String lastExecutionStatus;
        private String lastErrorMessage;
        
        public TaskExecutionInfo(String taskName, String description) {
            this.taskName = taskName;
            this.description = description;
            this.registeredAt = LocalDateTime.now();
            this.currentlyExecuting = false;
            this.lastExecutionStatus = "NEVER_EXECUTED";
        }
        
        // Getters and setters
        public String getTaskName() { return taskName; }
        public String getDescription() { return description; }
        public LocalDateTime getRegisteredAt() { return registeredAt; }
        public LocalDateTime getLastExecutionStart() { return lastExecutionStart; }
        public void setLastExecutionStart(LocalDateTime lastExecutionStart) { this.lastExecutionStart = lastExecutionStart; }
        public LocalDateTime getLastExecutionEnd() { return lastExecutionEnd; }
        public void setLastExecutionEnd(LocalDateTime lastExecutionEnd) { this.lastExecutionEnd = lastExecutionEnd; }
        public long getLastExecutionDuration() { return lastExecutionDuration; }
        public void setLastExecutionDuration(long lastExecutionDuration) { this.lastExecutionDuration = lastExecutionDuration; }
        public boolean isCurrentlyExecuting() { return currentlyExecuting; }
        public void setCurrentlyExecuting(boolean currentlyExecuting) { this.currentlyExecuting = currentlyExecuting; }
        public String getLastExecutionStatus() { return lastExecutionStatus; }
        public void setLastExecutionStatus(String lastExecutionStatus) { this.lastExecutionStatus = lastExecutionStatus; }
        public String getLastErrorMessage() { return lastErrorMessage; }
        public void setLastErrorMessage(String lastErrorMessage) { this.lastErrorMessage = lastErrorMessage; }
    }
    
    /**
     * Metrics for a scheduled task.
     */
    public static class TaskMetrics {
        private final AtomicLong executionCount = new AtomicLong(0);
        private final AtomicLong successCount = new AtomicLong(0);
        private final AtomicLong failureCount = new AtomicLong(0);
        private volatile long totalExecutionTime = 0;
        private volatile long maxExecutionTime = 0;
        
        public void incrementExecutionCount() {
            executionCount.incrementAndGet();
        }
        
        public void incrementSuccessCount() {
            successCount.incrementAndGet();
        }
        
        public void incrementFailureCount() {
            failureCount.incrementAndGet();
        }
        
        public synchronized void updateAverageExecutionTime(long executionTime) {
            totalExecutionTime += executionTime;
        }
        
        public synchronized void updateMaxExecutionTime(long executionTime) {
            if (executionTime > maxExecutionTime) {
                maxExecutionTime = executionTime;
            }
        }
        
        public long getExecutionCount() { return executionCount.get(); }
        public long getSuccessCount() { return successCount.get(); }
        public long getFailureCount() { return failureCount.get(); }
        public double getSuccessRate() {
            long total = executionCount.get();
            return total > 0 ? (double) successCount.get() / total * 100 : 0;
        }
        public synchronized long getAverageExecutionTime() {
            long total = executionCount.get();
            return total > 0 ? totalExecutionTime / total : 0;
        }
        public long getMaxExecutionTime() { return maxExecutionTime; }
    }
}

File: apps/api/src/jobs/schedulers/ScheduledTaskRegistry.ts
import { Logger } from 'winston';
import { createLogger } from '../../utils/logger';

/**
 * Registry service for managing and monitoring scheduled tasks.
 * Provides capabilities to track task execution, failures, and performance metrics.
 */
export class ScheduledTaskRegistry {
  private readonly logger: Logger;
  private readonly taskRegistry = new Map<string, TaskExecutionInfo>();
  private readonly taskMetrics = new Map<string, TaskMetrics>();

  constructor() {
    this.logger = createLogger('ScheduledTaskRegistry');
  }

  /**
   * Register a scheduled task for monitoring.
   */
  registerTask(taskName: string, description: string): void {
    const info = new TaskExecutionInfo(taskName, description);
    this.taskRegistry.set(taskName, info);
    this.taskMetrics.set(taskName, new TaskMetrics());
    
    this.logger.info(`Registered scheduled task: ${taskName} - ${description}`);
  }

  /**
   * Record the start of a task execution.
   */
  recordTaskStart(taskName: string): void {
    const info = this.taskRegistry.get(taskName);
    if (info) {
      info.setLastExecutionStart(new Date());
      info.setCurrentlyExecuting(true);
      
      const metrics = this.taskMetrics.get(taskName);
      if (metrics) {
        metrics.incrementExecutionCount();
      }
      
      this.logger.debug(`Task started: ${taskName}`);
    }
  }

  /**
   * Record the successful completion of a task execution.
   */
  recordTaskSuccess(taskName: string, executionTimeMs: number): void {
    const info = this.taskRegistry.get(taskName);
    if (info) {
      info.setLastExecutionEnd(new Date());
      info.setLastExecutionDuration(executionTimeMs);
      info.setCurrentlyExecuting(false);
      info.setLastExecutionStatus('SUCCESS');
      
      const metrics = this.taskMetrics.get(taskName);
      if (metrics) {
        metrics.incrementSuccessCount();
        metrics.updateAverageExecutionTime(executionTimeMs);
        metrics.updateMaxExecutionTime(executionTimeMs);
      }
      
      this.logger.debug(`Task completed successfully: ${taskName} (${executionTimeMs}ms)`);
    }
  }

  /**
   * Record a task execution failure.
   */
  recordTaskFailure(taskName: string, executionTimeMs: number, errorMessage: string): void {
    const info = this.taskRegistry.get(taskName);
    if (info) {
      info.setLastExecutionEnd(new Date());
      info.setLastExecutionDuration(executionTimeMs);
      info.setCurrentlyExecuting(false);
      info.setLastExecutionStatus('FAILED');
      info.setLastErrorMessage(errorMessage);
      
      const metrics = this.taskMetrics.get(taskName);
      if (metrics) {
        metrics.incrementFailureCount();
        metrics.updateAverageExecutionTime(executionTimeMs);
      }
      
      this.logger.warn(`Task failed: ${taskName} (${executionTimeMs}ms) - ${errorMessage}`);
    }
  }

  /**
   * Get execution information for a specific task.
   */
  getTaskInfo(taskName: string): TaskExecutionInfo | undefined {
    return this.taskRegistry.get(taskName);
  }

  /**
   * Get metrics for a specific task.
   */
  getTaskMetrics(taskName: string): TaskMetrics | undefined {
    return this.taskMetrics.get(taskName);
  }

  /**
   * Get all registered tasks.
   */
  getAllTasks(): Map<string, TaskExecutionInfo> {
    return new Map(this.taskRegistry);
  }

  /**
   * Get metrics for all tasks.
   */
  getAllMetrics(): Map<string, TaskMetrics> {
    return new Map(this.taskMetrics);
  }

  /**
   * Check if any tasks are currently executing.
   */
  hasRunningTasks(): boolean {
    return Array.from(this.taskRegistry.values()).some(info => info.isCurrentlyExecuting());
  }

  /**
   * Get count of currently executing tasks.
   */
  getRunningTaskCount(): number {
    return Array.from(this.taskRegistry.values()).filter(info => info.isCurrentlyExecuting()).length;
  }
}

/**
 * Information about a scheduled task execution.
 */
export class TaskExecutionInfo {
  private lastExecutionStart?: Date;
  private lastExecutionEnd?: Date;
  private lastExecutionDuration = 0;
  private currentlyExecuting = false;
  private lastExecutionStatus = 'NEVER_EXECUTED';
  private lastErrorMessage?: string;

  constructor(
    private readonly taskName: string,
    private readonly description: string,
    private readonly registeredAt = new Date()
  ) {}

  // Getters
  getTaskName(): string { return this.taskName; }
  getDescription(): string { return this.description; }
  getRegisteredAt(): Date { return this.registeredAt; }
  getLastExecutionStart(): Date | undefined { return this.lastExecutionStart; }
  getLastExecutionEnd(): Date | undefined { return this.lastExecutionEnd; }
  getLastExecutionDuration(): number { return this.lastExecutionDuration; }
  isCurrentlyExecuting(): boolean { return this.currentlyExecuting; }
  getLastExecutionStatus(): string { return this.lastExecutionStatus; }
  getLastErrorMessage(): string | undefined { return this.lastErrorMessage; }

  // Setters
  setLastExecutionStart(date: Date): void { this.lastExecutionStart = date; }
  setLastExecutionEnd(date: Date): void { this.lastExecutionEnd = date; }
  setLastExecutionDuration(duration: number): void { this.lastExecutionDuration = duration; }
  setCurrentlyExecuting(executing: boolean): void { this.currentlyExecuting = executing; }
  setLastExecutionStatus(status: string): void { this.lastExecutionStatus = status; }
  setLastErrorMessage(message: string): void { this.lastErrorMessage = message; }
}

/**
 * Metrics for a scheduled task.
 */
export class TaskMetrics {
  private executionCount = 0;
  private successCount = 0;
  private failureCount = 0;
  private totalExecutionTime = 0;
  private maxExecutionTime = 0;

  incrementExecutionCount(): void {
    this.executionCount++;
  }

  incrementSuccessCount(): void {
    this.successCount++;
  }

  incrementFailureCount(): void {
    this.failureCount++;
  }

  updateAverageExecutionTime(executionTime: number): void {
    this.totalExecutionTime += executionTime;
  }

  updateMaxExecutionTime(executionTime: number): void {
    if (executionTime > this.maxExecutionTime) {
      this.maxExecutionTime = executionTime;
    }
  }

  getExecutionCount(): number { return this.executionCount; }
  getSuccessCount(): number { return this.successCount; }
  getFailureCount(): number { return this.failureCount; }
  
  getSuccessRate(): number {
    return this.executionCount > 0 ? (this.successCount / this.executionCount) * 100 : 0;
  }
  
  getAverageExecutionTime(): number {
    return this.executionCount > 0 ? this.totalExecutionTime / this.executionCount : 0;
  }
  
  getMaxExecutionTime(): number { return this.maxExecutionTime; }
}